### PR TITLE
nixos/thanos: Added grpc-compression option

### PR DESCRIPTION
**Pull request by jpds**
## Description of changes

Expose `grpc-compression` option.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by jpds** [comment]
@anthonyroussel Hmm, I expected the bot to assign you as a reviewer

File: nixos/modules/services/monitoring/thanos.nix [hunk 1]
```diff
@@ -187,6 +187,10 @@ let
         Make sure this address is routable from other components.
       '';
 
+      grpc-compression = mkParam types.str ''
+        Compression algorithm to use for gRPC requests to other clients.
+      '';
+
```
Comment (anthonyroussel) [inline nixos/modules/services/monitoring/thanos.nix:193]: Hello, sorry for the late review.

https://github.com/search?q=repo%3Athanos-io%2Fthanos%20grpc-compression&type=code

* `grpc-compression` option is only available with `query` subcommand
* `receive` subcommand uses `receive.grpc-compression` option

I think you should move the `grpc-compression` parameter in the `query` parameters (~line 337).

What do you think?

**Comment by anthonyroussel** [comment]
I will take a look to the Thanos module soon to comply with [RFC 0042](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md#part-2).

We should migrate to use structural settings instead (same for Prometheus module).

File: nixos/modules/services/monitoring/thanos.nix [hunk 2]
```diff
@@ -642,6 +646,10 @@ let
 
     receive = params.common cfg.receive // params.objstore cfg.receive // {
 
+      grpc-compression = mkParam types.str ''
```
Comment (anthonyroussel) [inline nixos/modules/services/monitoring/thanos.nix:649]:
```suggestion
      receive.grpc-compression = mkParam types.str ''
```

https://github.com/thanos-io/thanos/blob/main/cmd/thanos/receive.go#L894

`receive` command uses `receive.grpc-compression` if I am not mistaken

File: nixos/modules/services/monitoring/thanos.nix [hunk 3]
```diff
@@ -642,6 +646,10 @@ let
...
+        Compression algorithm to use for gRPC requests to other clients.
```
Comment (anthonyroussel) [inline nixos/modules/services/monitoring/thanos.nix:650]:
```suggestion
        Compression algorithm to use for gRPC requests to other receivers.
```

To be consistent with the Thanos documentation on the query subcommand.
What do you think?

**Comment by anthonyroussel** [comment]
@ofborg test prometheus

**Review by anthonyroussel — APPROVED** [review]
Thanks! :+1:

**Review by Aleksanaa — APPROVED** [review]

**Pull request merged**